src/lib.rs

```diff
@@ -597,6 +597,190 @@ impl<A: Array> ArrayVec<A> {
+    /// assert_eq!(&vec[..], &[1, 2]);
+    /// ```
+    pub fn try_resize(&mut self, new_len: usize, value: A::Item)
+         -> Result<(), CapacityError>
+    {
+        let len = self.len();
+        if new_len > len {
+            self.extend_with(new_len - len, ExtendElement(value))
+        } else {
+            self.truncate(new_len);
+            Ok(())
+        }
+    }
+}
+
+impl<A: Array> ArrayVec<A>
+    where A::Item: Default,
+{
+    /// Resizes the `ArrayVec` in-place so that `len` is equal to `new_len`.
+    ///
+    /// Does the same thing as `try_resize_default`, but panics on error
+    /// instead of returning it as a `Result`.
+    pub fn resize_default(&mut self, new_len: usize) {
+        self.try_resize_default(new_len).unwrap();
+    }
+
+    /// Resizes the `ArrayVec` in-place so that `len` is equal to `new_len`.
+    ///
+    /// If `new_len` is greater than `len`, the `ArrayVec` is extended by the
+    /// difference, with each additional slot filled with `Default::default()`.
+    /// If `new_len` is less than `len`, the `ArrayVec` is simply truncated.
+    ///
+    /// This method uses `Default` to create new values on every push. If you'd
+    /// rather `Clone` a given value, use `try_resize`.
+    ///
+    /// If the capacity of the `ArrayVec` is smaller than the passed `new_len`,
+    /// an error is returned, and no changes to the `ArrayVec` is made.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use arrayvec::ArrayVec;
+    ///
+    /// let mut vec: ArrayVec<[_; 8]> = (1..4).collect();
+    /// assert!(vec.try_resize_default(5).is_ok());
+    /// assert_eq!(&vec[..], &[1, 2, 3, 0, 0]);
+    ///
+    /// let mut vec: ArrayVec<[_; 4]> = (1..5).collect();
+    /// assert!(vec.try_resize_default(2).is_ok());
+    /// assert_eq!(&vec[..], &[1, 2]);
+    /// ```
+    pub fn try_resize_default(&mut self, new_len: usize)
+        -> Result<(), CapacityError>
+    {
+        let len = self.len();
+        if new_len > len {
+            self.extend_with(new_len - len, ExtendDefault)
+        } else {
+            self.truncate(new_len);
+            Ok(())
+        }
+    }
+}
+
+trait ExtendWith<T> {
+    fn next(&self) -> T;
+    fn last(self) -> T;
+}
+
+struct ExtendElement<T>(T);
+impl<T: Clone> ExtendWith<T> for ExtendElement<T> {
+    fn next(&self) -> T { self.0.clone() }
+    fn last(self) -> T { self.0 }
+}
+
+struct ExtendDefault;
+impl<T: Default> ExtendWith<T> for ExtendDefault {
+    fn next(&self) -> T { Default::default() }
+    fn last(self) -> T { Default::default() }
+}
+
+impl<A: Array> ArrayVec<A> {
+    /// Extend the vector by `n` values, using the given generator.
+    ///
+    /// Doesn't extend the vector and returns an error if the number of added
+    /// elements exceed the capacity of the underlying array.
+    fn extend_with<E: ExtendWith<A::Item>>(&mut self, n: usize, value: E)
+        -> Result<(), CapacityError>
+    {
+        if self.len() + n > self.capacity() {
```

Comment:
This addition feels wrong, as if it could wrap around. But it's a private function and all callers are without problem.